cryptol-remote-api/python/cryptol/commands.py

```diff
@@ -161,14 +183,62 @@ def __init__(self, connection : HasProtocolState, expr : Any) -> None:
     def process_result(self, res : Any) -> Any:
         return res['type schema']
 
+
 class SmtQueryType(str, Enum):
     PROVE = 'prove'
     SAFE  = 'safe'
     SAT   = 'sat'
 
-class CryptolProveSat(argo.Command):
+class SmtResult(ABC):
+    """Abstract class for SMT query results."""
+    qtype: SmtQueryType
+    
+    def __init__(self, qtype : SmtQueryType) -> None:
+        if qtype in [SmtQueryType.PROVE, SmtQueryType.SAFE, SmtQueryType.SAT]:
+            self.qtype = qtype
+        else:
+            raise ValueError("Unknown SMT query type: " + qtype)            
+
+class SmtUnsatResult(SmtResult):
+    def __init__(self, qtype : SmtQueryType) -> None:
+        super(SmtUnsatResult, self).__init__(qtype)
+
+    def __bool__(self) -> bool:
+        """An unsat result is a success for PROVE and SAFE queries but a
+        failure for SAT queries. Thus, the object is truthy iff qtype != SAT.
+        """
+        return self.qtype != SmtQueryType.SAT
+
+class SmtSatResult(List[CryptolPython], SmtResult):
+    def __init__(self, qtype : SmtQueryType, cxs : List[CryptolPython]) -> None:
+        SmtResult.__init__(self, qtype)
+        list.__init__(self, cxs)
+
+class SmtInvalidResult(List[CryptolPython], SmtResult):
+    def __init__(self, qtype : SmtQueryType, cxs : List[CryptolPython]) -> None:
+        SmtResult.__init__(self, qtype)
+        list.__init__(self, cxs)
+
+SmtQueryResult = Union[SmtResult, OfflineSmtQuery]
```

Comment:
Tell me what you think! It seems like having to worry about whether or not we got an `OfflineSmtQuery` in every instance where we call one of these functions is silly, because we can _only_ get those under certain circumstances. I.e., when the solver is _specified_ as an offline solver. Would it be nice to have `prove_offline`, `sat_offline`, and `check_offline` which _only_ returns a `OfflineSmtQuery`?

We could then be clearer about our SMT `Solver` type (I forget what it's called as I'm typing this), and have one type for the online solvers (probably called whatever it is now), and then have a _specific_ `OfflineSolver` type that delineates the available offline solvers. Then the online and offline versions of the functions could be specific about what solvers they allow, and the return types could not worry about the impossible result types! (i.e., online query will never return `OfflineSmtQuery`, etc).

Aye? Thoughts?

Reply:
In the spirit of having the Python interface mirror the REPL interface, I didn't make separate methods for offline queries, I instead used `@overloads` to give the two cases (the `OnlineSolver` and `OfflineSolver` cases, as I called them) distinct types. Let me know what you think